### ISSUE TITLE
Elliptic Curve cryptography for public key authentication

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5350]
|<img alt="termim's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| termim reported|
|-|-|
|Trac ID|trac#5350|
|Type|defect|
|Created|2011-11-01 18:33:43Z|
|Branch|https://github.com/twisted/twisted/tree/ecdsa-keys-5350-1|

Once I connect to some host that supplies its public key in ecdsa I can't use twisted's conch.ssh anymore. It always raises BadKeyError:

```

2011-11-01 13:35:46-0400 [-] Log opened.
2011-11-01 13:35:46-0400 [-] Starting factory <twisted.conch.client.direct.SSHClientFactory instance at 0x105bf80>
2011-11-01 13:35:46-0400 [SSHClientTransport,client] kex alg, key alg: diffie-hellman-group-exchange-sha1 ssh-rsa
2011-11-01 13:35:46-0400 [SSHClientTransport,client] outgoing: aes256-cbc hmac-sha1 none
2011-11-01 13:35:46-0400 [SSHClientTransport,client] incoming: aes256-cbc hmac-sha1 none
2011-11-01 13:35:47-0400 [SSHClientTransport,client] Unhandled Error
	Traceback (most recent call last):
	  File "/usr/lib64/python2.7/site-packages/twisted/python/log.py", line 84, in callWithLogger
	    return callWithContext({"system": lp}, func, *args, **kw)
	  File "/usr/lib64/python2.7/site-packages/twisted/python/log.py", line 69, in callWithContext
	    return context.call({ILogContext: newCtx}, func, *args, **kw)
	  File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
	    return self.currentContext().callWithContext(ctx, func, *args, **kw)
	  File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
	    return func(*args,**kw)
	--- <exception caught here> ---
	  File "/usr/lib64/python2.7/site-packages/twisted/internet/selectreactor.py", line 146, in _doReadOrWrite
	    why = getattr(selectable, method)()
	  File "/usr/lib64/python2.7/site-packages/twisted/internet/tcp.py", line 460, in doRead
	    rval = self.protocol.dataReceived(data)
	  File "/usr/lib64/python2.7/site-packages/twisted/conch/ssh/transport.py", line 438, in dataReceived
	    self.dispatchMessage(messageNum, packet[1:])
	  File "/usr/lib64/python2.7/site-packages/twisted/conch/ssh/transport.py", line 453, in dispatchMessage
	    f(payload)
	  File "/usr/lib64/python2.7/site-packages/twisted/conch/ssh/transport.py", line 1182, in ssh_KEX_DH_GEX_REPLY
	    d = self.verifyHostKey(pubKey, fingerprint)
	  File "/usr/lib64/python2.7/site-packages/twisted/conch/client/direct.py", line 87, in verifyHostKey
	    fingerprint)
	  File "/usr/lib64/python2.7/site-packages/twisted/conch/client/default.py", line 74, in verifyHostKey
	    or os.path.expanduser("~/.ssh/known_hosts")
	  File "/usr/lib64/python2.7/site-packages/twisted/conch/client/knownhosts.py", line 417, in fromPath
	    entry = HashedEntry.fromString(line)
	  File "/usr/lib64/python2.7/site-packages/twisted/conch/client/knownhosts.py", line 253, in fromString
	    stuff, keyType, key, comment = _extractCommon(string)
	  File "/usr/lib64/python2.7/site-packages/twisted/conch/client/knownhosts.py", line 59, in _extractCommon
	    key = Key.fromString(keyString.decode('base64'))
	  File "/usr/lib64/python2.7/site-packages/twisted/conch/ssh/keys.py", line 81, in fromString
	    raise BadKeyError('cannot guess the type of %r' % data)
	twisted.conch.ssh.keys.BadKeyError: cannot guess the type of '\x00\x00\x00\x13ecdsa-sha2-nistp256\x00\x00\x00\x08nistp256\x00\x00\x00A\x04j\x82Z\x88v\xad\xf9\xa3\xc4\x94\x19\xff\xc5\x81\x0e6\x7f\xaf\xce\xf5\xbbC\x1d\x8f\x99l\x99=\x95\xc9\x1f\x1e{\xb0\xb7:Z\x9fh7\x91\xa8lf\x03\xc2`x\x05\x1a\xadqV\x0f|\xa7 `2\x80\xda\x0f\x8e\x11'
	
2011-11-01 13:35:47-0400 [SSHClientTransport,client] connection lost
2011-11-01 13:35:47-0400 [SSHClientTransport,client] Stopping factory <twisted.conch.client.direct.SSHClientFactory instance at 0x105bf80>
```

Attachments:

* [ecdsa.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b2a/b2adc33af0589fb4ae76a6a8303e87068cc01ad6/520f0de11e0264efc0db7736cce6e08dc3c83cf2.py) (332 bytes) - added by thijs on 2012-03-14 19:02:58Z - parse ECDSA
* [id_ecdsa_conch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b2a/b2adc33af0589fb4ae76a6a8303e87068cc01ad6/a369fbd2584ebd7ad4a3bea9a9e60dfe40e3496d) (444 bytes) - added by thijs on 2012-07-04 09:19:10Z - private
* [id_ecdsa_conch.pub](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b2a/b2adc33af0589fb4ae76a6a8303e87068cc01ad6/b2a6bc672d589c998d77b941a146fe5d8b74f41a.pub) (265 bytes) - added by thijs on 2012-07-04 09:19:44Z - public

<details><summary>Searchable metadata</summary>

```
trac-id__5350 5350
type__defect defect
reporter__termim termim
priority__highest highest
milestone__ 
branch__branches_ecdsa_keys_5350_1 branches/ecdsa-keys-5350-1
branch_author__z3p z3p
status__closed closed
resolution__duplicate duplicate
component__conch conch
keywords__ssh ssh
time__1320172423000000 1320172423000000
changetime__1478775550156898 1478775550156898
version__None None
owner__z3p z3p
cc__z3p cc__thijs cc__mmoya
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

The _regular-releases_ milestone is for release automation related tasks.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli commented|
|-|-|

I filed [#5530](https://github.com/twisted/twisted/issues/5530) recently, which may be related (or possibly a duplicate)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

Attached test script generates a `ValueError` on an ECDSA private key:

```
$ ssh-keygen -b 521 -t ecdsa -f id_ecdsa_conch -N qwerty
Generating public/private ecdsa key pair.
Your identification has been saved in id_ecdsa_conch.
Your public key has been saved in id_ecdsa_conch.pub.
The key fingerprint is:
9c:36:91:fd:d5:f9:ee:92:55:65:49:3d:0a:d6:2a:8c 

The key's randomart image is:
+--[ECDSA  521]---+
|            . ..o|
|         o o . +=|
|        = o o oo+|
|       E = o o  o|
|        S . .   o|
|       . .     ..|
|               o.|
|              o. |
|               ..|
+-----------------+

$ python ecdsa.py id_ecdsa_conch qwerty
```

```python
Traceback (most recent call last):
  File "ecdsa.py", line 13, in <module>
    key = Key.fromFile(fn, passphrase=phrase)
  File "/home/thijs/workspaces/opensource/software/twisted/svn/Twisted/trunk/twisted/conch/ssh/keys.py", line 62, in fromFile
    return Class.fromString(file(filename, 'rb').read(), type, passphrase)
  File "/home/thijs/workspaces/opensource/software/twisted/svn/Twisted/trunk/twisted/conch/ssh/keys.py", line 90, in fromString
    return method(data, passphrase)
  File "/home/thijs/workspaces/opensource/software/twisted/svn/Twisted/trunk/twisted/conch/ssh/keys.py", line 216, in _fromString_PRIVATE_OPENSSH
    keyData = DES3.new(decKey, DES3.MODE_CBC, iv).decrypt(b64Data)
ValueError: IV must be 8 bytes long
```

Replying to [cyli](#note_3):
> I filed [#5530](https://github.com/twisted/twisted/issues/5530) recently, which may be related (or possibly a duplicate)

Can you try the attached script with your key?

**Comment by twisted-trac** [comment]
|<img alt="termim's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|termim commented|
|-|-|

Actually there are two bugs here
 1. missing support for ecdsa keys
 2. bug when reading knownhosts file
For the second a simpler one case I've opened another ticket [#5616](https://github.com/twisted/twisted/issues/5616) (with a patch attached).

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|z3p commented|
|-|-|

Can you attach an ECDSA key? I couldn't create one on OS X or on Debian.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra set owner to z3p|
|-|-|

Attached the keys, generated like described in [comment:4 comment 4].

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|z3p commented|
|-|-|

(In [34683]) Branching to 'ecdsa-keys-5350'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|z3p commented|
|-|-|

(In [34684]) add ECDSA key guessing

Refs [#5350](https://github.com/twisted/twisted/issues/5350)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|z3p set status to `assigned`|
|-|-|

Unfortunately at the moment, PyCrypto doesn't support ECDSA keys.  I'm going to do some reading, though, and look into implementing it.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|z3p commented|
|-|-|

Does this key have a passphrase I can use? Alternatively, can you generate one which isn't encrypted?

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|z3p commented|
|-|-|

Nevermind, I'm an idiot. It's in the command you generated it with.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|z3p commented|
|-|-|

(In [34703]) Branching to 'ecdsa-keys-5350-1'

**Comment by twisted-trac** [comment]
|<img alt="kalou's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|kalou commented|
|-|-|

Hello, given the age of this report, would it be possible to catch this Exception in twisted.conch.ssh.userauth so that we refuse the key ?

This would prevent an unsupported key proposal to break the negociation process and proceed further to password auth, waiting for a better fix. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

> Hello, given the age of this report, would it be possible to catch this Exception in twisted.conch.ssh.userauth so that we refuse the key ?
> This would prevent an unsupported key proposal to break the negociation process and proceed further to password auth, waiting for a better fix. Thanks!

Already done, I think: [#5616](https://github.com/twisted/twisted/issues/5616).  If you still have this problem with Twisted 12.1 or newer, please file a ticket explaining what's up.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_17"></a>|@adiroiban set status to `closed`|
|-|-|

So this should be fixed by [#5616](https://github.com/twisted/twisted/issues/5616)

Also as of [#8828](https://github.com/twisted/twisted/issues/8828), ECDSA key can not be read/loaded.

Please reopen if you still have problems.
